README.md
# ZeRO: Zero-Knowledge Risk Oracle" 


CHANGES.txt
v1.0.0, April 23 2019 -- Initial Release


example/tests/runtest.sh
#!/bin/bash

rm *.txt
cd ..


./zero.py -d tests/exD1.dx -n `seq 20 1 150` -B "tests/res1.txt"
./zero.py -d tests/exD2.dx -n `seq 20 1 150` -B "tests/res2.txt"
./zero.py -d tests/exD3.dx -n `seq 20 1 150` -B "tests/res3.txt"
./zero.py -d tests/exD4.dx -n `seq 20 1 150` -B "tests/res4.txt"
./zero.py -d tests/exN1.dx -n `seq 20 1 150` -B "tests/ren1.txt"
./zero.py -d tests/exN2.dx -n `seq 20 1 150` -B "tests/ren2.txt"
./zero.py -d tests/exN3.dx -n `seq 20 1 150` -B "tests/ren3.txt"
./zero.py -d tests/exN4.dx -n `seq 20 1 150` -B "tests/ren4.txt"

cd tests/

sed -i 's/ /\n/3; P; D' res1.txt
sed -i 's/ /\n/3; P; D' res2.txt
sed -i 's/ /\n/3; P; D' res3.txt
sed -i 's/ /\n/3; P; D' res4.txt
sed -i 's/ /\n/3; P; D' ren1.txt
sed -i 's/ /\n/3; P; D' ren2.txt
sed -i 's/ /\n/3; P; D' ren3.txt
sed -i 's/ /\n/3; P; D' ren4.txt

gnuplot gplot.gnu


ehrzero/example/tests/run_test.sh
#!/bin/bash

FILES='exD1.dx  exD2.dx  exD3.dx  exD4.dx  exN1.dx  exN2.dx  exN3.dx  exN4.dx'
rm -rf *txt

#for i in {10..150}
#do
#    for fl in $FILES
#    do
#	./zero.py -d $fl -n $i >> "$fl"'.txt'
#    done
#done
cd ..
PROG="./zero.py "
parallel -j 10 $PROG -d ./tests/{1} -n {2} -B ./tests/{1}.txt ::: $FILES ::: `seq 10 150`


ehrzero/tuning.py
import numpy as np
import pandas as pd
import numpy as np
from random import randint, choice
from random import uniform as randfloat
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
import lightgbm as lgb
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import Normalizer

def split_combine(df, ratio = 0.81):
    X = df.drop("target", 1)
    y = df.target
    Xtt, Xtv, ytt, ytv = train_test_split(X, y, test_size=0.81, random_state=1321)
    TRA = Xtt
    TRA["target"] = ytt
    TES = Xtv
    TES["target"] = ytv
    return TRA, TES

def class_weights(arr):
    dict(pd.Series([i[0] for i in arr]).value_counts(normalize = True))
"""
params = {'boosting_type': 'gbdt',
          'max_depth' : randint(222, 330),
          'objective': 'binary',
          'nthread': 28, # Updated from nthread
          'num_leaves': randint(80, 125),
          'learning_rate': randfloat(0.22,0.25),
          'max_bin': randint(50, 450),
          'subsample_for_bin': randint(8000, 10500), # 
          'subsample': randfloat(0.9925, 1),
          'subsample_freq': randint(35, 45),
          'colsample_bytree': randfloat(0.5, 0.57),
          'reg_alpha': randfloat(7, 9),
          'reg_lambda': randfloat(8, 14),
          'min_split_gain': 0.0012,#randfloat(0.001, 0.003),  #
          'min_child_weight': randfloat(6, 8), #
          'min_child_samples': randint(450, 530),
          'scale_pos_weight': randfloat(0.62, 0.72),
          'min_data_in_leaf': randint(100, 500),
          'metric' : 'auc',
          'early_stopping_round' : 37,
          'verbosity' : 0}
"""
    
def RUN_XG(train_set, test_set, k_splits):
    hyperparameters = {'boosting_type': 'gbdt',
          'max_depth' : randint(222, 330),
          'objective': 'binary',
          'nthread': 28, # Updated from nthread
          'num_leaves': randint(80, 125),
          'learning_rate': randfloat(0.22,0.25),
          'max_bin': randint(50, 450),
          'subsample_for_bin': randint(8000, 10500), # 
          'subsample': randfloat(0.9925, 1),
          'subsample_freq': randint(35, 45),
          'colsample_bytree': randfloat(0.5, 0.57),
          'reg_alpha': randfloat(7, 9),
          'reg_lambda': randfloat(8, 14),
          'min_split_gain': 0.0012,#randfloat(0.001, 0.003),  #
          'min_child_weight': randfloat(6, 8), #
          'min_child_samples': randint(450, 530),
          'scale_pos_weight': randfloat(0.62, 0.72),
          'min_data_in_leaf': randint(100, 500),
          'metric' : 'auc',
          'early_stopping_round' : 37,
          'verbosity' : 0}
    auc_scores = []
    DELELIST = ['target']
    TOTAL = pd.concat([train_set, test_set])
    X = TOTAL.drop(DELELIST, 1)
    y = TOTAL.target
    
    skf = StratifiedKFold(n_splits=k_splits)
    skf.get_n_splits(X, y)
    
    for train_index, test_index in skf.split(X, y):
        Xtrain, Xtest = X.iloc[train_index], X.iloc[test_index]
        ytrain, ytest = y.iloc[train_index], y.iloc[test_index]
        Xt, Xv, yt, yv = train_test_split(Xtrain, ytrain, test_size=0.38)
        Xt_array = Xt.values
        yt_array = yt.values.astype('int').flatten()
        Xv_array = Xv.values
        yv_array = yv.values.astype('int').flatten()
        
        clf = lgb.LGBMClassifier(**hyperparameters)
        clf.fit(Xt_array, yt_array, 
                eval_set = [(Xt_array, yt_array), (Xv_array, yv_array)],
                eval_metric = 'auc',
                early_stopping_rounds = 37, 
                verbose = False)
        y_predict = clf.predict_proba(Xtest, num_iteration=clf.best_iteration_)
        auc = roc_auc_score(ytest, y_predict[:,1])
        auc_scores.append(auc)
        
    output = hyperparameters.copy()
    output["AUC"] = np.mean(auc_scores)
    print(" XXXXX X X >>", np.mean(auc_scores))
    return output

def RUN_NB(train_set, test_set, k_splits):
    hyperparameters = {"var_smoothing": randfloat(0.01,5)}
    auc_scores = []
    DELELIST = ['target']
    ftest = test_set
    Xtest = ftest.drop(DELELIST, 1)
    ytest = ftest.target
    for i in range(k_splits):
        ftrain = train_set
        X = ftrain.drop(DELELIST, 1)
        y = ftrain.target
        scaler = MinMaxScaler()
        scaler.fit(X)
        del ftrain
        clf = GaussianNB(**hyperparameters)
        clf.fit(scaler.transform(X) + 10, y)
        y_predict = clf.predict_proba(scaler.transform(Xtest) + 10)
        auc = roc_auc_score(ytest, y_predict[:,1])
        auc_scores.append(auc)
        
    output = hyperparameters.copy()
    output["AUC"] = np.mean(auc_scores)
    print(" XXXXX X X >>", np.mean(auc_scores))
    return output

def random_search(num_iters, train_set, test_set, model, k_splits = 5, downsample = 0):
    learning_results = []
    for i in range(num_iters):
        if i % 14 == 0:
            print(" {  %d  }  " % i)
        if model == "LGBM":
            learning_results.append(RUN_XG(train_set, test_set, k_splits))
        elif model == "NB":
            learning_results.append(RUN_NB(train_set, test_set, k_splits))
        else:
            print("Model not recognized.\nCurrently available models:\n'LGBM' : LightGBM\n'NB' : GaussianNaiveBayes\n")
            return
    DF = pd.DataFrame(learning_results)
    print("======\nMEAN ::: ",np.mean(DF.AUC))
    return DF


ehrzero/dataviz.py
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
import pandas as pd
import numpy as np
import itertools
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score

def plot_confusion_matrix(df, classes = ["0", "1"],
           normalize=False,
           title='Confusion matrix',
           cmap=plt.cm.Blues):
    """
    This function prints and plots the confusion matrix.
    Normalization can be applied by setting `normalize=True`.
    """
    print("F1 score      >> ", sklearn.metrics.f1_score(df["truth"], df["pred_label"]))
    print("ROC AUC score >> ", roc_auc_score(df["truth"], df["pred_label"]))
    cm = sklearn.metrics.confusion_matrix(df["truth"], df["pred_label"])
    if normalize:
        cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
    plt.imshow(cm, interpolation='nearest', cmap=cmap)
    plt.title(title)
    plt.colorbar()
    tick_marks = np.arange(len(classes))
    plt.xticks(tick_marks, classes, rotation=45)
    plt.yticks(tick_marks, classes)
    fmt = '.2f' if normalize else 'd'
    thresh = cm.max() / 2.
    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
        plt.text(j, i, format(cm[i, j], fmt),
                 horizontalalignment="center",
                 color="white" if cm[i, j] > thresh else "black")

    plt.tight_layout()
    plt.ylabel('True label')
    
def analyze_class_predictions(df,thr):
    """
        Plot the classwise prediction distributions
    """
    fig, (ax1, ax2) = plt.subplots(nrows=2, sharey=True ,figsize=(10,12))
    sns.distplot(df[df.truth == 0]["predicted_risk"], bins = 30, color = "navy", ax = ax1)
    #ax1.axvline(thr, 'k')
    ax1.set_title("Risks predicted for Negative records")
    sns.distplot(df[df.truth == 1]["predicted_risk"], bins = 30, color = "maroon", ax = ax2)
    ax2.set_title("Risks predicted for Positive records")
    #ax2.axvline(thr, 'k')
    plt.show()
    
def plot_tuning_stats(df, metric = "AUC", min_value_to_show = 0):
    if min_value_to_show:
        df = df[df[metric] > min_value_to_show]
    plt.figure(figsize=(6,6))
    plt.style.use('seaborn-whitegrid')
    FEATURES = list(df.columns)[1:]
    for feat in FEATURES:
        plt.figure(figsize=(5,5))
        plt.plot(df[feat], df[metric], '*')
        plt.title(feat)
        plt.show()

ehrzero/etiology.py








































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































































import pandas as pd
import numpy as np

def count_codes(DATA, KEYS):
    counts = []
    for ID in DATA.patient_id:
        ID = ID[:-1]
        if not ID in patients.keys():
            counts.append(0)
        else:
            try:
                i = 0
                for disease in patients[ID]:
                    disease = (disease[:4] + "0").ljust(6," ")[:-3]
                    if disease in list(KEYS.code):
                        i += 1
                counts.append(i)
            except:
                counts.append(0)
    return counts

def retrieve_diagnoses_before_x(x, records, TARGET_CODES):
    aut_week = 0
    filtered_records = []
    # Get all the disease records that happened prior to (x+1)th week
    for record in records.split(":")[1:]:
        if "|" in record:
            vals = record.split("|")
            if vals[1].strip() in TARGET_CODES:
                if not aut_week:
                    aut_week = int(vals[0].strip())
                else:
                    aut_week = min(aut_week, int(vals[0].strip()))
            elif int(vals[0].strip()) <= x:
                filtered_records.append(record)
    if aut_week:
        filtered_records = [i for i in filtered_records if int(i.split("|")[0].strip()) < x]
    return [i.split("|")[1].strip() for i in filtered_records]


def retrieve_disease_stats(DF, codedict, patients, categories, offset = 0):
    done = False
    STATS = {}
    DESCR = {}
    GROUP = {}
    COUNTS = {}
    io = 0
    for ID in list(DF.patient_id):
        try:
            if ID in patients.keys():
                already = []
                io += 1
                for disease in patients[ID]:
                    if disease[3] == "." and disease[4] == " ":
                        if not done:
                            done = True
                        if offset:
                            disease = (disease[:4] + "0").ljust(6," ")[:-offset]
                        else:
                             disease = (disease[:4] + "0").ljust(6," ")
                    if ":" in disease:
                        if offset:
                            disease = disease.split(":")[0].ljust(6," ")[:-offset]
                        else:
                            disease = disease.split(":")[0].ljust(6," ")
                    if offset:
                        if len(disease) > 3:
                            disease = disease[:3]
                    DESCR[disease] = codedict[disease] if disease in codedict.keys() else 'NaN'
                    GROUP[disease] = catt[disease] if disease in categories.keys() else 'NaN'
                    if disease not in already:
                        if disease in STATS.keys():
                            STATS[disease] += 1
                            already.append(disease)
                        elif len(disease) > 2:
                            STATS[disease] = 1
                            already.append(disease)
                        else:
                            print(disease)
        except:
            continue
    print(len(STATS.keys()))
    print(len(DESCR.keys()))
    print(len(GROUP.keys()))
    print(len(STATS.keys()))
    df = pd.DataFrame.from_dict({
                                 'code': list(STATS.keys()), 
                                 'description': list(DESCR.values()), 
                                 'disease_group': list(GROUP.values()),
                                 'occurrences': list(STATS.values())
                                })
    return df

feature_engineering.py
import numpy as np
import pandas as pd
import itertools
from sklearn.metrics import roc_auc_score, roc_curve, auc

def augment_agg(X):
    mean = np.array(X.mean(axis = 1)).reshape(X.shape[0],1)
    std = np.array(X.std(axis = 1)).reshape(X.shape[0],1)
    rang = np.array((X.max(axis = 1) - X.min(axis = 1))).reshape(X.shape[0],1)

    X = np.append(X,mean,1)
    X = np.append(X,std,1)
    X = np.append(X,rang,1)
    
    return X

def longest_one_streak(lst):
    return max(sum(1 for x in l if x == 1) for n, l in itertools.groupby(lst))

def optimal_cutoff(labels, preds):
    ####################################
    # The optimal cut off would be where tpr is high and fpr is low
    # tpr - (1-fpr) is zero or near to zero is the optimal cut off point
    ####################################
    fpr, tpr, cutoff = roc_curve(labels, preds)
    roc_auc = auc(fpr, tpr)
    #print("Area under the ROC curve : %f" % roc_auc)
    i = np.arange(len(tpr)) # index for df
    roc = pd.DataFrame({'fpr' : pd.Series(fpr, index=i),'tpr' : pd.Series(tpr, index = i), '1-fpr' : pd.Series(1-fpr, index = i), 'tf' : pd.Series(tpr - (1-fpr), index = i), 'thresholds' : pd.Series(cutoff, index = i)})
    return float(roc.ix[(roc.tf-0).abs().argsort()[:1]]["thresholds"])

def get_dynamics(VALUES, TEST_SEQ_LENGTH):
                        first_half = pd.Series([np.mean(i[:int(TEST_SEQ_LENGTH/2)]) for i in VALUES])
                        second_half = pd.Series([np.mean(i[int(TEST_SEQ_LENGTH/2):TEST_SEQ_LENGTH]) for i in VALUES])
                        dynamics = (second_half/first_half).fillna(0)
                        dynamics[np.isinf(dynamics)] = 0
                        return dynamics

def get_max_streak_length(arr, x): 
    # intitialize count 
    count = 0 
    # initialize max 
    result = 0 
    for i in arr: 
        # Reset count when 0 is found 
        if (i == x): 
            count += 1
        # If 1 is found, increment count 
        # and update result if count  
        # becomes more. 
        else: 
            count = 0
        result = max(result, count)  
    return result  

ehrzero/preprocessing.py
import pandas as pd
import numpy as np
from random import randint
from math import floor
import os

#############################################
##  Functions for ICD-9 to CSV conversion  ##
#############################################

def files(path):  
    for file in os.listdir(path):
        if os.path.isfile(os.path.join(path, file)):
            yield file

def optimize_memory(df):
    for col in df.columns:
        if df[col].dtype != object:

            # make variables for Int, max and min
            IsInt = False
            mx = df[col].max()
            mn = df[col].min()
            # Integer does not support NA, therefore, NA needs to be filled
            if not np.isfinite(df[col]).all(): 
                #NAlist.append(col)
                df[col].fillna(mn-1,inplace=True)  
            # test if column can be converted to an integer
            asint = df[col].fillna(0).astype(np.int64)
            result = (df[col] - asint)
            result = result.sum()
            if result > -0.01 and result < 0.01:
                IsInt = True

            # Make Integer/unsigned Integer datatypes
            if IsInt:
                if mn >= 0:
                    if mx < 255:
                        df[col] = df[col].astype(np.uint8)
                    elif mx < 65535:
                        df[col] = df[col].astype(np.uint16)
                    elif mx < 4294967295:
                        df[col] = df[col].astype(np.uint32)
                    else:
                        df[col] = df[col].astype(np.uint64)
                else:
                    if mn > np.iinfo(np.int8).min and mx < np.iinfo(np.int8).max:
                        df[col] = df[col].astype(np.int8)
                    elif mn > np.iinfo(np.int16).min and mx < np.iinfo(np.int16).max:
                        df[col] = df[col].astype(np.int16)
                    elif mn > np.iinfo(np.int32).min and mx < np.iinfo(np.int32).max:
                        df[col] = df[col].astype(np.int32)
                    elif mn > np.iinfo(np.int64).min and mx < np.iinfo(np.int64).max:
                        df[col] = df[col].astype(np.int64)    

            # Make float datatypes 32 bit
            else:
                df[col] = df[col].astype(np.float32)
    return df
        
def get_csv_dataset(DATAPATH, pattern, target = 'target', separate_x_y = True):
    data_dict = {'target': [], 'county': [], 'record': []}
    count = 0
    for i, filepath in enumerate(files(DATAPATH)):
        if pattern in filepath:
            with open(os.path.join(DATAPATH, filepath), 'r') as file:
                lines = file.readlines()
                for record_line in lines:
                        count += 1
                        data_dict['target'].append(int(filepath[1:4] == "POS"))
                        data_dict['county'].append(filepath[4:])
                        data_dict['record'].append(record_line)
    df = pd.DataFrame(data_dict)
    if not separate_x_y:
        return df
    else:
        return df.drop(target, 1), df[target]
    
def get_csv_dataset_integrated(DATAPATH):   
    data_dict = {'patient_id': [], 'gender': [], 'fips': [], 'record': [], 'target': []}
    with open(DATAPATH, 'r') as f:
        ao = f.readlines()
        for i, record in enumerate(ao):
            data = record.split()
            data_dict['patient_id'].append((data[1]))
            data_dict['gender'].append(data[0][0])
            data_dict['fips'].append(data[0][4:])
            data_dict['record'].append(" ".join(data[2:]))
            data_dict['target'].append(int(data[0][1:4] == "POS"))
    df = pd.DataFrame(data_dict)
    return df

def intc(i):
    try:
        return int(i)
    except:
        return 0 
    
def retrieve_raw_records(PATH, FIRST_N_WEEKS, full = False, integrated = False):
    if full:
        if integrated:
            test = get_csv_dataset_integrated(PATH)
        else:
            test = get_csv_dataset(PATH, 'M', separate_x_y = False)
    else:
        test = pd.read_csv('TEST_SET.csv')
    test['sequence'] = [[intc(i) for i in record[:-2].split(' ')][:FIRST_N_WEEKS] for record in test.record] #intc(i) to get integer

    X = pd.DataFrame(test.sequence.tolist(), columns=['SEQ_%i' % i for i in range(FIRST_N_WEEKS)])
    X['county'] = test.fips
    X['target'] = test.target
    X['gender'] = test.gender
    if integrated:
        X['patient_id'] = test.patient_id
    X.dropna(axis=0, inplace = True)
    return X


def optimal_cutoff(labels, preds):
    ####################################
    # Return the optimal binary cutoff value
    # given the ROC curve made by input labels and predictions
    # --------------------------------------------------------
    # The optimal cut off would be where tpr is high and fpr is low
    # tpr - (1-fpr) is zero or near to zero is the optimal cut off point
    ####################################
    fpr, tpr, cutoff = roc_curve(labels, preds)
    roc_auc = auc(fpr, tpr)
    i = np.arange(len(tpr)) # index for df
    roc = pd.DataFrame({'fpr' : pd.Series(fpr, index=i),'tpr' : pd.Series(tpr, index = i), '1-fpr' : pd.Series(1-fpr, index = i), 'tf' : pd.Series(tpr - (1-fpr), index = i), 'thresholds' : pd.Series(cutoff, index = i)})
    return float(roc.ix[(roc.tf-0).abs().argsort()[:1]]["thresholds"])

def augment_agg(X):
    mean = X.mean(axis = 1).reshape(X.shape[0],1)
    std = X.std(axis = 1).reshape(X.shape[0],1)
    rang = (X.max(axis = 1) - X.min(axis = 1)).reshape(X.shape[0],1)
    X = np.append(X,mean,1)
    X = np.append(X,std,1)
    X = np.append(X,rang,1)
    return X

def convert_to_dx(custom_id,seq,sep='|',delim=',',YR='2019'):
    '''
        Converts user input to DX format
        DX format: 000003878|  4|U|F|26099|2007^034.0 :  5:276|034.0 :...
        seq: input string
        sep: separator of sub-items within a record
        delim: separator for records
        Thus, for sep='|',delim=','
        a valid user input is:
        amy|M|01003,21|766.00,29|566.0,36|V0.20
        For sep=' ' and delim ':', the same input:
        amy M 01003:21 766.00:29 566.0:36 V0.20
        
        We can skip the FIPS (then 00000 is assumed),
        and we can skip the name/id, 
        then default SUBJECT is assumed
        returns: DX format input
    ''' 
    x=seq.split(delim)
    C=x[1:]
    
    hdr=x[0].split(sep)
    if len(hdr)==3:
        ID=hdr[0]
        G=hdr[1]
        FIPS=hdr[2]
    if len(hdr)==2:
        ID=hdr[0]
        G=hdr[1]
        FIPS='00000'
    if len(hdr)==1:
        G=hdr[0]
        ID = str(custom_id).rjust(9,"0")#str(randint(1,999999999)).rjust(9,"0")#'000000199'
        FIPS='00000'
    C_age_code = [(int(x.split(sep)[0]),
                 floor(int(x.split(sep)[0]) / 52 + 1),
                 x.split(sep)[1]) for x in C ]
    min_age = np.array([x[1] for x in C_age_code]).min()
    HDR = ID + "|  " + str(min_age) + '|U|' + G + "|" + FIPS + '|' + YR + '^'
    return (HDR+'|'.join([x[2]+':  '+str(x[1])+':'+str(x[0])  for x in C_age_code]) + "|").replace("\n", "") + '\r\n'

def files(path):  
    for file in os.listdir(path):
        if os.path.isfile(os.path.join(path, file)):
            yield file

def reduce_memory(df):
    for col in df.columns:
        if df[col].dtype != object:
            # make variables for Int, max and min
            IsInt = False
            mx = df[col].max()
            mn = df[col].min()
            # Integer does not support NA, therefore, NA needs to be filled
            if not np.isfinite(df[col]).all(): 
                #NAlist.append(col)
                df[col].fillna(mn-1,inplace=True)  
            # test if column can be converted to an integer
            asint = df[col].fillna(0).astype(np.int64)
            result = (df[col] - asint)
            result = result.sum()
            if result > -0.01 and result < 0.01:
                IsInt = True
            # Make Integer/unsigned Integer datatypes
            if IsInt:
                if mn >= 0:
                    if mx < 255:
                        df[col] = df[col].astype(np.uint8)
                    elif mx < 65535:
                        df[col] = df[col].astype(np.uint16)
                    elif mx < 4294967295:
                        df[col] = df[col].astype(np.uint32)
                    else:
                        df[col] = df[col].astype(np.uint64)
                else:
                    if mn > np.iinfo(np.int8).min and mx < np.iinfo(np.int8).max:
                        df[col] = df[col].astype(np.int8)
                    elif mn > np.iinfo(np.int16).min and mx < np.iinfo(np.int16).max:
                        df[col] = df[col].astype(np.int16)
                    elif mn > np.iinfo(np.int32).min and mx < np.iinfo(np.int32).max:
                        df[col] = df[col].astype(np.int32)
                    elif mn > np.iinfo(np.int64).min and mx < np.iinfo(np.int64).max:
                        df[col] = df[col].astype(np.int64)    
            # Make float datatypes 32 bit
            else:
                df[col] = df[col].astype(np.float32)
    return df
        
def get_csv_dataset(DATAPATH, pattern, target = 'target', separate_x_y = True):
    data_dict = {'target': [], 'county': [], 'record': []}
    count = 0
    for i, filepath in enumerate(files(DATAPATH)):
        if pattern in filepath:
            with open(os.path.join(DATAPATH, filepath), 'r') as file:
                lines = file.readlines()
                for record_line in lines:
                        count += 1
                        data_dict['target'].append(int(filepath[1:4] == "POS"))
                        data_dict['county'].append(filepath[4:])
                        data_dict['record'].append(record_line)
    df = pd.DataFrame(data_dict)
    if not separate_x_y:
        return df
    else:
        return df.drop(target, 1), df[target]
    
def get_csv_dataset_integrated(DATAPATH, WEEKS = 0):   
    data_dict = {'patient_id': [], 'gender': [], 'fips': [], 'record': [], 'target': []}
    with open(DATAPATH, 'r') as f:
        ao = f.readlines()
        for i, record in enumerate(ao):
            if i % 1000000 == 0 and i != 0:
                print(i)
            data = record.split()
            data_dict['patient_id'].append(("A" + data[1]))
            data_dict['gender'].append(data[0][0])
            data_dict['fips'].append(data[0][4:])
            if WEEKS:
                data_dict['record'].append(" ".join(data[2:(WEEKS + 2)]))
            else:
                data_dict['record'].append(" ".join(data[2:]))
            data_dict['target'].append(int(data[0][1:4] == "POS"))
    df = pd.DataFrame(data_dict)
    return df

def intc(i):
    try:
        return int(i)
    except:
        return 0 
    
def retrieve_raw_records(PATH, FIRST_N_WEEKS, full = False, all_rows = False, integrated = False):
    if full:
        data = get_csv_dataset_integrated(PATH)
        return data[["patient_id", "gender", "fips", "record", "target"]]
    else:
        if all_rows:
            if integrated:
                data = get_csv_dataset_integrated(PATH, FIRST_N_WEEKS)
            else:
                data = get_csv_dataset(PATH, 'M', separate_x_y = False)
        else:
            test = pd.read_csv('TEST_SET.csv')
            test['sequence'] = [[intc(i) for i in record[:-2].split(' ')][:FIRST_N_WEEKS] for record in test.record] 
            X = pd.DataFrame(test.sequence.tolist(), columns=['SEQ_%i' % i for i in range(FIRST_N_WEEKS)])
            X['county'] = test.fips
            X['target'] = test.target
            X['gender'] = test.gender
            if integrated:
                X['patient_id'] = test.patient_id
            return test
        #data.dropna(axis=0, inplace = True)
        return data.dropna(axis=0)[["patient_id", "gender", "fips", "record", "target"]]

def optimal_cutoff(labels, preds):
    ####################################
    # Return the optimal binary cutoff value
    # given the ROC curve made by input labels and predictions
    # --------------------------------------------------------
    # The optimal cut off would be where tpr is high and fpr is low
    # tpr - (1-fpr) is zero or near to zero is the optimal cut off point
    ####################################
    fpr, tpr, cutoff = roc_curve(labels, preds)
    roc_auc = auc(fpr, tpr)
    #print("Area under the ROC curve : %f" % roc_auc)
    i = np.arange(len(tpr)) # index for df
    roc = pd.DataFrame({'fpr' : pd.Series(fpr, index=i),'tpr' : pd.Series(tpr, index = i), '1-fpr' : pd.Series(1-fpr, index = i), 'tf' : pd.Series(tpr - (1-fpr), index = i), 'thresholds' : pd.Series(cutoff, index = i)})
    return float(roc.ix[(roc.tf-0).abs().argsort()[:1]]["thresholds"])

def retrieve_diagnoses_before_x(x, records):
    aut_week = 0
    filtered_records = []
    # Get all the disease records that happened prior to (x+1)th week
    for record in records.split(":")[1:]:
        if "|" in record:
            vals = record.split("|")
            if vals[1].strip() in AUTISTIC:
                if not aut_week:
                    aut_week = int(vals[0].strip())
                else:
                    aut_week = min(aut_week, int(vals[0].strip()))
            elif int(vals[0].strip()) <= 150:
                filtered_records.append(record)
    if aut_week:
        filtered_records = [i for i in filtered_records if int(i.split("|")[0].strip()) < aut_week]
    return [i.split("|")[1].strip() for i in filtered_records]

def three_split(df, train_r=.6, proc_r=.2, seed=None):
    np.random.seed(seed)
    perm = np.random.permutation(df.index)
    m = len(df.index)
    train_end = int(train_r * m)
    validate_end = int(proc_r * m) + train_end
    train = df.ix[perm[:train_end]]
    processing = df.ix[perm[train_end:validate_end]]
    test = df.ix[perm[validate_end:]]
    return train, processing, test

def get_data(conditions, path, gender, fips, jump):
    CONT = sp.county_cluster(fips, jump)
    dfs = []
    #KEYS = []
    train_sets = {}
    processing_sets = {}
    test_sets = {}
    for KEY in conditions:
        print(KEY)
        df = pd.read_csv(path % (gender, KEY))
        df = df[df.county.isin(CONT)]
        splits = three_split(df,
                            train_r = .4,
                            proc_r = .4)
        train_sets[KEY] = splits[0]
        processing_sets[KEY] = splits[1]
        test_sets[KEY] = splits[2]
    print(train_sets['Infectious_Disease'].shape)
    print(processing_sets['Infectious_Disease'].shape)
    print(test_sets['Infectious_Disease'].shape)
    return train_sets, processing_sets, test_sets

def build_pfsas(datasets, LLK):
    pfsas = []
    for key in datasets.keys():
        print(key)
        X = datasets[key].drop(['county'], 1)
        y = datasets[key].target
        Z=Z3Classifier(result_path = "RESULT_PATH", llk_path = LLK)
        #X['target'] = list(y)
        #print(">>", test.shape[0])
        Z.fit(X) # fits with BOTH feats and labels
        pfsas.append(Z)
    return pfsas

def obtain_lls(datasets, pfsas):
    dfs = []
    for key, Z in zip(datasets.keys(), pfsas):
        try:
            preds = {}
            test = datasets[key].drop(['county', 'target'], 1)
            target = datasets[key].target
            LL = Z.predict_loglike(test) # raw outpuit is zip
            preds[key + "_NEG"] = list(LL[0])
            preds[key + "_POS"] = list(LL[1])
            if len(preds[key + "_NEG"]) != test.shape[0] or len(preds[key + "_POS"]) != test.shape[0]:
                print("[ - - - ] Output size mismatch, retrying")
                continue
            df = pd.DataFrame(preds)
            df['patient_id'] = test['patient_id']
            df['target'] = list(target) #### ## #
            dfs.append(df)
            print(key)
        except:
            print('ERROR >> %s' % key)
            continue
            
    df = reduce(lambda left, right: pd.merge(left, right, on='patient_id',
                                                    how='outer').drop_duplicates(), dfs)
    print("SDDDDDDD")
    cols = list(df.columns)
    num = 0
    for i in range(len(cols)):
        if "target" in cols[i]:
            cols[i] = "target_%d" % num
            num += 1
    df.columns = cols
    print('KEOEK')
    targets = [df[var] for var in list(df.columns) if "target" in var]
    print("IKDPODK")
    full_target = reduce(lambda left, right : left.combine_first(right), targets)
    df.drop([i for i in list(df.columns) if "target" in i], 1, inplace = True)
    print('reduced')
    df['target'] = full_target
    if 'Infectious_Disease_NEG' in df.columns:
        df = df.dropna(subset=['Infectious_Disease_NEG'])
    return df

